### DISCUSSION TITLE
Using next-urql with getServerSideProps and client side urql integration

### DISCUSSION THREAD

**Discussion by kennetpostigo** [urql-graphql/urql#2488] [Q&A] [answered]
Hi,

I have a few questions regarding workflows and how to keep a globally available `urql` client in sync with a server-side instance of `urql`.

I use `urql` in a few different ways but mainly:
  1. In `getServerSideProps` to run queries and mutations

  2. In my `pages/_app.tsx` I have my application wrapped in the `<Provider />` exported from `urql` and passing a new instance of a client to it. I do this so that my xstate machines that are in React.context and consumed via hooks can access an instance of the client with the `useClient` hook. I then pass actions that use the `client` to make queries/mutations when configuring the machines.
  
How I can update the `client` instance passed to the `<Provider />` after queries/mutations are ran in `getServerSideProps`. In [this example](https://github.com/FormidableLabs/urql/blob/main/examples/with-next/pages/server.js) I see that query is ran in `getServerSideProps` it passes down `urqlState` and then within the component `useQuery` is used, and I assume that query simply reads from the cache and returns the data. How does that work for mutations and `useMutation` and will that update the `urql` client instance that was passed to `Provider` from the `urql` package?

Lastly, I use cookie based authentication, and i'm curious how to keep both client/server instance of the `urql` client aware of the cookie with `authExchange` so that when server finishes, the client picks up with the cookie where the server left off?

**Comment by JoviDeCroock** [comment, marked as answer]
[This is the code in next-urql](https://github.com/FormidableLabs/urql/blob/main/packages/next-urql/src/with-urql-client.ts#L63-L75) that keeps the state from the server on the client. When passing down props from `getServerSideProps` we can pick these up on the client and leverage them to restore data in our `ssrExchange` meaning the data we fetched server-side is now synchronously available to the client-side.

For the `authExchange` part you will need to switch between req.cookie (which sends down a `Set-Cookie` header) and document.cookie, this means that you have to create the app with `req` in closure.
